packages/reactive-element/src/decorators/query-assigned-elements.ts

```diff
@@ -0,0 +1,60 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/*
+ * IMPORTANT: For compatibility with tsickle and the Closure JS compiler, all
+ * property decorators (but not class decorators) in this file that have
+ * an @ExportDecoratedItems annotation must be defined as a regular function,
+ * not an arrow function.
+ */
+
+import {ReactiveElement} from '../reactive-element.js';
+import {decorateProperty} from './base.js';
+
+/**
+ * A property decorator that converts a class property into a getter that
+ * returns the `assignedElements` of the given `slot`. Provides a declarative
+ * way to use
+ * [`slot.assignedElements`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLSlotElement/assignedElements).
+ *
+ * @param slotName A string name of the slot. Leave empty or use `''` for
+ *     default slot.
+ * @param options Object that sets options for nodes to be returned. See
+ *     [MDN parameters section](https://developer.mozilla.org/en-US/docs/Web/API/HTMLSlotElement/assignedElements#parameters)
+ *     for available options.
+ *
+ * ```ts
+ * class MyElement {
+ *   @queryAssignedElements('list')
+ *   listItems;
+ *
+ *   render() {
+ *     return html`
+ *       <slot name="list"></slot>
+ *     `;
+ *   }
+ * }
+ * ```
+ * @category Decorator
+ */
+export function queryAssignedElements(
+  slotName = '',
+  options?: AssignedNodesOptions | undefined
```

Comment:
Two things...
1. I wonder if we should fix the queryAssignedNodes API and take an options bag here? That way it's as easy to specify AssignedNodesOptions without a slot name as it is to specify a slot name without options.
2. Since queryAssignedNodes takes a selector to filter elements, should this?

Reply:
Yeah I agree that we may as well make this more useful, especially as the other decorator provides a filter. Below are two API options. Do you have a preference or suggestion?

### Option 1

```ts
interface QueryAssignedElementsOptions {
  /**
   * Name of the slot. Leave undefined for unnamed default slot.
   */
  slotName?: string,
  /**
   * If it exists, is passed directly into `HTMLSlotElement.assignedElements(options)`.
   */
  options?: AssignedNodesOptions,
  /**
   * A string which filters the results to elements that match the given css selector.
   */
  selector?: string
}

function queryAssignedElements(
  options?: QueryAssignedElementsOptions
) { ... }
```

### Option 2

Flatten the options bag (but there could be collisions in the future if more options than `flatten` are added), i.e.: `interface QueryAssignedElementsOptions extends AssignedNodesOptions { /* slotName, selector */ }`



Does it also make sense to migrate other decorators with or shortly after this one with the options bag approach for consistency, or do we need to wait for a certain release?

Thank you!

Reply:
I like option 2.

I think it'd be ideal to update `queryAssignedNodes` as well, but in a non-breaking way by type-checking the first argument and adding a TS override.

Reply:
Updated this commit. Also created a follow up task for `queryAssignedNodes`: https://github.com/lit/lit/issues/2332